### PR TITLE
[grid-ui] Re-route all internal path with subpath

### LINKED ISSUE #11647: [🚀 Feature]: Allow user to provide a different base URL for the grid UI
### Feature and motivation

For the moment the base URL of a grid is locked to be /ui/ at root level.
For my usage i had to manually change the path in the file : selenium/grid-ui/package.json at line 38 `"homepage": "/ui"` to `"homepage": "/selenium/ui"`

Because for my usage the base URL host multiple services (so i have my URL like this www.myUrl.com/selenium/ui) 

Currently i found no other way to provide the grid a custom base without having to recompile everything after a manual change.

### Usage example

The grid could simply read an environment variable such as GRID_BASE_URL and if this variable exist use it instead of the /ui written in the code

### PR DESCRIPTION
**Pull request by yashcho**
[grid-ui] Re-route all internal path with subpath

Fixes #11647

**Thanks for contributing to Selenium!**
**A PR well described will help maintainers to quickly review and merge it**

Before submitting your PR, please check our [contributing](https://github.com/SeleniumHQ/selenium/blob/trunk/CONTRIBUTING.md) guidelines.
Avoid large PRs, help reviewers by making them as simple and short as possible.

### Description
Grid UI can now re-route all internal router request for hub that contains /ui and /graphql paths through provided subpath

### Types of changes

- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

### Checklist

- [x] I have read the [contributing](https://github.com/SeleniumHQ/selenium/blob/trunk/CONTRIBUTING.md) document.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [ ] I have added tests to cover my changes.
- [ ] All new and existing tests passed.

### CODE REVIEW

**Comment by yashcho** [comment]
I tested the jar directly in my setup environment were selenium is hosted under a context / subpath path but upstream points to the root path of application dashboard.

Only there are some warnings which are popping up in developer tools console don't know why, below is the screenshot

image

Any help to solve this will be helpful 😄 ...

**Comment by yashcho** [comment]
@diemol and @krmahadevan any update on this?

**Comment by krmahadevan** [comment]
@yashcho - I haven't had time to get to this. A bit swamped with stuff at work. Apologies. I believe @diemol is busy with the Selenium Conference. Please give some time. I will get to this asap.

**Comment by yashcho** [comment]
Thanks @krmahadevan for the quick response.

File: java/src/org/openqa/selenium/grid/web/PrefixAwareJarFileResource.java [hunk 1]
```diff
@@ -0,0 +1,80 @@
+// Licensed to the Software Freedom Conservancy (SFC) under one
```
Comment (krmahadevan) [inline java/src/org/openqa/selenium/grid/web/PrefixAwareJarFileResource.java:1]: This class duplicates stuff. Instead can we please consider refactoring this class into something like this ?

```java
package org.openqa.selenium.grid.web;

import java.util.Arrays;
import java.util.List;
import java.util.Optional;
import java.util.jar.JarFile;

public class PrefixAwareJarFileResource extends JarFileResource {
    private final String subPath;

    public PrefixAwareJarFileResource(JarFile jarFile, String entryName, String stripPrefix, String subPath) {
        super(jarFile, entryName, stripPrefix);
        this.subPath = subPath;
    }

  @Override
  protected JarFileResource newInstance(JarFile jarFile, String entryName, String prefix) {
    return new PrefixAwareJarFileResource(jarFile, entryName, prefix, subPath);
  }

  private static final List<String> EXCLUSION_LIST = Arrays.asList(
    ".woff", ".woff2", ".svg", ".png"
  );

    @Override
    public Optional<byte[]> read() {
      Optional<byte[]> readByteArray = super.read();
      boolean found = EXCLUSION_LIST.parallelStream()
        .anyMatch(entryName::endsWith);
      if ( found) {
        return readByteArray;
      }
      if (!readByteArray.isPresent()) {
        return readByteArray;
      }
      String readString = new String(readByteArray.get());

      // All file such as index.html, *.js, *.css and *.js.map
      readString = readString.replaceAll("/ui/", subPath.concat("/ui/"))
        .replaceAll("\\+\"/graphql\"", "+\"".concat(subPath).concat("/graphql\""));

      return Optional.of(readString.getBytes());
    }
}
```

File: java/src/org/openqa/selenium/grid/web/JarFileResource.java [hunk 2]
```diff
@@ -32,9 +32,9 @@
 
 public class JarFileResource implements Resource {
```
Comment (krmahadevan) [inline java/src/org/openqa/selenium/grid/web/JarFileResource.java:33]: Enhance this class to include a new method that looks like below

```java
  protected JarFileResource newInstance(JarFile jarFile, String entryName, String prefix) {
    return new JarFileResource(jarFile, entryName, prefix);
  }

```

And this method would be used as below in this class

```java
  @Override
  public Optional<Resource> get(String path) {
    Require.nonNull("Path", path);

    if (!isDirectory()) {
      return Optional.empty();
    }

    String name = stripPrefix + stripLeadingSlash(path);

    ZipEntry entry = jarFile.getEntry(name);
    return Optional.ofNullable(entry).map(e -> newInstance(jarFile, entry.getName(), name));
  }
```

```java
  @Override
  public Set<Resource> list() {
    if (!isDirectory()) {
      return ImmutableSet.of();
    }

    String prefix = entryName.endsWith("/") ? entryName : entryName + "/";
    int count = prefix.split("/").length + 1;

    return jarFile.stream()
      .filter(e -> e.getName().startsWith(prefix))
      .filter(e -> !e.getName().equals(entryName))
      .filter(e -> !e.getName().equals(prefix))
      .filter(e -> e.getName().split("/").length == count)
      .map(e -> newInstance(jarFile, e.getName(), prefix))
      .collect(ImmutableSet.toImmutableSet());
  }
```

File: rust/Cargo.Bazel.lock
Comment (krmahadevan) [inline rust/Cargo.Bazel.lock:1]: Can we please revert this change, since it does not have anything to do with the current changeset

File: rust/Cargo.lock
Comment (krmahadevan) [inline rust/Cargo.lock:1]: Can we please revert this change, since it does not have anything to do with the current changeset

File: java/test/org/openqa/selenium/grid/gridui/ReverseProxyOverallGridTest.java [hunk 3]
```diff
@@ -0,0 +1,141 @@
...
+import static org.junit.jupiter.api.Assertions.assertEquals;
+import static org.openqa.selenium.grid.gridui.Urls.whereIs;
+import static org.openqa.selenium.json.Json.MAP_TYPE;
+import static org.openqa.selenium.remote.http.HttpMethod.GET;
+import static org.openqa.selenium.support.ui.ExpectedConditions.textToBe;
+import static org.openqa.selenium.support.ui.ExpectedConditions.visibilityOfAllElementsLocatedBy;
+import static org.openqa.selenium.support.ui.ExpectedConditions.visibilityOfElementLocated;
+import static org.openqa.selenium.testing.Safely.safelyCall;
+
+class ReverseProxyOverallGridTest {
```
Comment (krmahadevan) [inline java/test/org/openqa/selenium/grid/gridui/ReverseProxyOverallGridTest.java:59]: Can we also please add a test for the redirection working fine for graphql endpoints ?

**Review by krmahadevan — CHANGES_REQUESTED** [review]
@yashcho - I have left some comments. Please help take a look.

**Comment by krmahadevan** [comment]
@yashcho 

> Only there are some warnings which are popping up in developer tools console don't know why, below is the screenshot

It looks like this is due to the redirecting not working. Selenium has a default `index.html` instead of throwing a 404.

More context on this is available in [this stackoverflow](https://stackoverflow.com/a/35038344)

**Review by diemol — COMMENTED** [review]
Thank you for reviewing, @krmahadevan!

File: java/src/org/openqa/selenium/grid/web/JarFileResource.java (hunk 2 above)
Comment (yashcho) [reply java/src/org/openqa/selenium/grid/web/JarFileResource.java:33]: Done the changes.

File: java/src/org/openqa/selenium/grid/web/PrefixAwareJarFileResource.java (hunk 1 above)
Comment (yashcho) [reply java/src/org/openqa/selenium/grid/web/PrefixAwareJarFileResource.java:1]: Done the changes.

File: rust/Cargo.Bazel.lock
Comment (yashcho) [reply rust/Cargo.Bazel.lock:1]: Yes but deleting the file is leading to not build the jar anymore in my development environment. How should I proceed?

File: rust/Cargo.lock
Comment (yashcho) [reply rust/Cargo.lock:1]: Yes but deleting the file is leading to not build the jar anymore in my development environment. How should I proceed?

File: java/test/org/openqa/selenium/grid/gridui/ReverseProxyOverallGridTest.java (hunk 3 above)
Comment (yashcho) [reply java/test/org/openqa/selenium/grid/gridui/ReverseProxyOverallGridTest.java:59]: What should a test for graphql do?

I mean creating a test case is ok but what should it check, because graphql endpoint works with post request so I might have to compare the result with in-memory values somewhat like mocking the graphql endpoint. Do you need me to build such test case?

**Comment by yashcho** [comment]
> @yashcho
> 
> > Only there are some warnings which are popping up in developer tools console don't know why, below is the screenshot
> 
> It looks like this is due to the redirecting not working. Selenium has a default `index.html` instead of throwing a 404.
> 
> More context on this is available in [this stackoverflow](https://stackoverflow.com/a/35038344)

It seems the changes from your end resolved it.

**Comment by yashcho** [comment]
@krmahadevan I have restructure the code as per the suggestion just need some answers for questions I raised to complete other suggestions.. 😊

File: java/test/org/openqa/selenium/grid/gridui/ReverseProxyOverallGridTest.java (hunk 3 above)
Comment (yashcho) [reply java/test/org/openqa/selenium/grid/gridui/ReverseProxyOverallGridTest.java:59]: Got the file with already existing test cases so created new file named "ReverseProxyGraphqlHandlerTest" and added the subpath changes.

Tested both the files with command :

`bazel test --cache_test_results=no java/test/org/openqa/selenium/grid/gridui:large-tests`

`bazel test --cache_test_results=no java/test/org/openqa/selenium/grid/graphql:small-tests`

**Comment by yashcho** [comment]
> @krmahadevan I have restructure the code as per the suggestion just need some answers for questions I raised to complete other suggestions.. 😊

Sorry @krmahadevan it seems the rebasing went wrong but my changes are added and reflected too, can you please verify and fix this PR?

**Comment by krmahadevan** [comment]
> > @krmahadevan I have restructure the code as per the suggestion just need some answers for questions I raised to complete other suggestions.. 😊
> 
> Sorry @krmahadevan it seems the rebasing went wrong but my changes are added and reflected too, can you please verify and fix this PR?

@yashcho - I would suggest that you do the following

1. Create a new branch off of `trunk`
2. Cherry pick your commit into that new branch
3. Amend the commit if needed to remove the rust related changes (They are not required for your PR and so I dont think those changes should be there in the commit)
4. Add the test for graphql (you need to ensure that the url with the modified base path is responding to queries). You dont need to be adding all the tests that test out each and every functionality/capability of the graphql exposed endpoints.

**Comment by yashcho** [comment]
Thank you @krmahadevan I have complete till step 3, for step 4 I am confused still with statement **"responding to queries"** can you elaborate more on this or provide an example?, so I will structure it and add it to branch and send a new PR.

**Comment by krmahadevan** [comment]
@yashcho what i meant was is that you only need to add a test that shows one http get call ( say for eg we want to retrieve the status from graphql ) works fine when we specify a new base path prefix. 

I think the codebase already has elaborate tests which ensures that all graphql functionalities are working fine. We dont need to be duplicating all those tests again. We only want to check if graphql endpoint works fine when we add a base path as prefix.

**Comment by yashcho** [comment]
So @krmahadevan if I understand well, you want to say 
- Do GET request against graphql endpoint with subpath using driver.get
- Fetch the HTTP status code
- Then check with assert.Equals to 200 code.

If this the case, I face issue to get 200 if I perform GET to graphql endpoint as graphql works with POST request method only it seems else please correct my understanding points.

Reference :- https://www.selenium.dev/documentation/grid/advanced_features/graphql_support/

**Comment by krmahadevan** [comment]
> So @krmahadevan if I understand well, you want to say
> 
> * Do GET request against graphql endpoint with subpath using driver.get
> * Fetch the HTTP status code
> * Then check with assert.Equals to 200 code.
> 
> If this the case, I face issue to get 200 if I perform GET to graphql endpoint as graphql works with POST request method only it seems else please correct my understanding points.
> 
> Reference :- https://www.selenium.dev/documentation/grid/advanced_features/graphql_support/

@yashcho - We seem to be going around in circles with this.

There is currently a test class `org.openqa.selenium.grid.graphql.GraphqlHandlerTest` that has elaborate tests for all the endpoints support that the graphql integration in Selenium Grid offers.

My ask is that you add an additional test method  something similar to `org.openqa.selenium.grid.graphql.GraphqlHandlerTest#shouldBeAbleToGetGridUri` but which uses an explicitly set base path, and which will test that the graphql functionalities work fine.


That way, we can rely on `org.openqa.selenium.grid.graphql.GraphqlHandlerTest#GraphqlHandlerTest` for all the functional testing of graphql support, including the working when a custom base path has been specified

I hope that clears the ambiguities and helps you move forward.

**Comment by yashcho** [comment]
So @krmahadevan I have created new MR (https://github.com/SeleniumHQ/selenium/pull/11886) should I close this one?

**Comment by yashcho** [comment]
Any suggestion for above question @krmahadevan ?

**Comment by krmahadevan** [comment]
> So @krmahadevan I have created new MR (#11886) should I close this one?

@yashcho - Yes. Please have this closed.
